OWML.ModHelper.Input/ModInputHandler.cs

```diff
@@ -0,0 +1,457 @@
+        }
+
+        private bool IsNewlyReleased_Single(IModInputCombination combination, bool keep = false)
+        {
+            return (!IsPressed_Single(combination)) && combination.IsFirst(keep) && (Time.realtimeSinceStartup - combination.GetLastPressedMoment() < _tapKeep);
+        }
+
+        public bool IsPressed(IModInputCombination combination)
+        {
+            return IsPressed_Combo(combination) || IsPressed_Single(combination);
+        }
+
+        public bool IsNewlyPressed(IModInputCombination combination, bool keep = false)
+        {
+            return IsPressed(combination) && combination.IsFirst(keep);
+        }
+
+        public bool WasTapped(IModInputCombination combination)
+        {
+            return (!(IsPressed_Combo(combination) || IsPressed_Single(combination))) && (Time.realtimeSinceStartup - combination.GetLastPressedMoment() < _tapKeep) && (combination.GetPressDuration() < _tapDuration);
+        }
+
+        public bool WasNewlyReleased(IModInputCombination combination, bool keep = false)
+        {
+            return (!(IsPressed_Combo(combination) || IsPressed_Single(combination))) && (Time.realtimeSinceStartup - combination.GetLastPressedMoment() < _tapKeep);
+        }
+
+        private Int64 ParseCombination(string combo, bool forRemoval = false)
+        {
+            combo = combo.Trim().Replace("ctrl", "control");
+            int[] thisCombination = new int[_maxComboLength];
+            int i = 0;
+            foreach (string key in combo.Split('+'))
+            {
+                if (i >= _maxComboLength)
+                    return (int)RegistrationCode.CombinationTooLong;
+                KeyCode code;
+                if (key.Contains("xbox_"))
+                {
+                    string xboxKey = key.Substring(5);
+                    var xboxCode = (XboxButton)Enum.Parse(typeof(XboxButton), xboxKey, true);
+                    if (Enum.IsDefined(typeof(XboxButton), xboxCode))
+                    {
+                        code = InputTranslator.GetKeyCode(xboxCode, false);
+                    }
+                    else
+                    {
+                        return (int)RegistrationCode.InvalidCombination;
+                    }
+                }
+                else
+                {
+                    string changedKey = key;
+                    if (key == "control")
+                    {
+                        changedKey = "leftcontrol";
+                    }
+                    else if (key == "shift")
+                    {
+                        changedKey = "leftshift";
+                    }
+                    else if (key == "alt")
+                    {
+                        changedKey = "leftalt";
+                    }
+                    code = (KeyCode)Enum.Parse(typeof(KeyCode), changedKey, true);
```

Comment:
I'd make a separate method for this. `string` as input, `KeyCode` as output. Which means for the special cases you can directly return KeyCode.LeftControl, for instance, instead of using strings.

Splitting long methods into smaller methods like this, even when they aren't gonna be reused, is a good way to increase readability without relying on comments. Maybe you've already done that a bit here, but just in case :)

Reply:
actually... that's a good idea

Reply:
done, maybe will change scope from private to public in future